system/macros/pagelist.php

```diff
@@ -0,0 +1,219 @@
+		$pages = $this->paginatePages($pages);
+
+		return $this->renderPages($pages);
+	}
+
+	private function paginatePages(array $pages) {
+		// Allow limiting the number of pages (for now just by a
+		// macro option, in the future maybe by e.g. GET
+		// parameters too).
+		// TODO: Better checking of option values
+		$limit = intval($this->macro_tag->getAttribute("limit"));
+		if ($limit)
+			return array_slice($pages, 0, $limit);
+
+		return $pages;
+	}
+
+	private function sortPages($pages) {
+		$keyfuncs = [
+			'date' => function($page) {
+				$date = $page->getSortDateTime();
+				return ($date ? $date->getTimestamp() : 0);
+			},
+			'title' => function($page) {
+				return $page->getTitle();
+			}
+		];
+
+		$sortattr = $this->macro_tag->getAttribute('sort');
+		if ($sortattr) {
+			$order = SORT_ASC;
+			if ($sortattr[0] == '-') {
+				$order = SORT_DESC;
+				$sortattr = substr($sortattr, 1);
+			}
+
+			if (!array_key_exists($sortattr, $keyfuncs))
+				throw new UnexpectedValueException("Invalid sort key: $sortattr. Valid options are: " . implode(', ', array_keys($keyfuncs)));
+			$keyfunc = $keyfuncs[$sortattr];
+
+			// This does a parallel sorting of the $keys and $pages
+			// arrays, using $keys as the sort value (only looking
+			// at $pages if two pages have the same key).
+			$keys = array_map($keyfunc, $pages);
+			array_multisort($keys, $order,  SORT_NATURAL | SORT_FLAG_CASE, $pages);
+		}
+		return $pages;
+	}
+
+	private function lookupTags($tagsattr) {
+		$tags = [];
+		if ($tagsattr) {
+			foreach (explode(',', $tagsattr) as $single) {
+				$split = explode('/', $single, 2);
+				if (count($split) == 1) {
+					$lang = $this->O_O->getContentLanguage();
+					$label = $split[0];
+				} else {
+					list($lang, $label) = $split;
+				}
+				$tag = HyphaTags::findTagByLabel($lang, $label);
+				if (!$tag)
+					throw new UnexpectedValueException("Invalid tag in language $lang: $label");
+				$tags[] = $tag;
+			}
+		}
+		return $tags;
+	}
+
+	private function getPages() {
+		// Allow filtering by tag
+		$tagsattr = $this->macro_tag->getAttribute("tags");
+		$tags = self::lookupTags($tagsattr);
+
+		$excludeTagsattr = $this->macro_tag->getAttribute("exclude-tags");
+		$excludeTags = $this->lookupTags($excludeTagsattr);
+
+		// Include private pages only when logged in and enabled
+		// TODO: Better checking of option values (and maybe
+		// allow Yes/true/TrUe/etc.)
+		$includePrivate = isUser() && ($this->macro_tag->getAttribute("include-private") == "yes");
+
+		// Allow filtering by page type
+		$pageTypes = [];
+		$pageTypesAttr = $this->macro_tag->getAttribute("pagetypes");
+		if ($pageTypesAttr) {
+			$pageTypes = explode(',', $pageTypesAttr);
+			$validTypes = hypha_getDataTypes();
+			foreach ($pageTypes as $type) {
+				if (!array_key_exists($type, $validTypes))
+					throw new UnexpectedValueException("Invalid page type: $type. Valid options are: " . implode(', ', array_keys($validTypes)));
```

Comment:
Do we want to abort or fail with a message? If in the future one page type is removed/changed then the entire macro becomes broken.

Reply:
This exception is caught further up, so it's just the macro that fails, not the entire page. That seems to me the safest and clearest way to fail, you could ignore the missing value and render something with a warning on top, but then it's really easy to miss the warning, so I think I'd prefer to just have the macro fail completely.